add percentage of valid pixels to metadata when computing raster band statistics

metzm opened the pull request:
Raster bands may contain pixels encoded as nodata. The proposed change adds a new STATISTICS_VALID_PERCENT metadata item to a raster band when raster band statistics are computed. Useful to get an estimate on the percentage of valid (not nodata) pixels, equivalent to e.g. percent cloud cover.

rouault commented on gdal/gcore/gdalrasterband.cpp:
```diff
@@ -3938,9 +3939,10 @@ static void ComputeStatisticsInternalGeneric( int nXCheck,
             {
                 const int iOffset = iX + iY * nBlockXSize;
                 const GUInt32 nValue = pData[iOffset];
+                nSampleCount ++;
```
could be moved out of the loops as nSampleCount += nXCheck * nYCheck

rouault commented on gdal/gcore/gdalrasterband.cpp:
```diff
@@ -5219,6 +5279,34 @@ GDALRasterBand::ComputeStatistics( int bApproxOK,
         SetStatistics( dfMin, dfMax, dfMean, dfStdDev );
     }
 
+    if( nValidCount == 0 )
```
as I see the below snippet is found before, can we factor it in a helper function ?

rouault commented:
```
Checking for tabulation characters...
gcore/gdalrasterband.cpp
FAIL: tabulations detected. Please remove them!
```

CI tests that need to be updated:
```
  TEST: stats_approx_stats_flag ... {'STATISTICS_MAXIMUM': '0', 'STATISTICS_MINIMUM': '0', 'STATISTICS_STDDEV': '0', 'STATISTICS_VALID_PERCENT': '100', 'STATISTICS_MEAN': '0', 'STATISTICS_APPROXIMATE': 'YES'}
fail
    line 820: did not get expected metadata
  TEST: stats_approx_stats_flag_float ... {'STATISTICS_MAXIMUM': '0', 'STATISTICS_MINIMUM': '0', 'STATISTICS_STDDEV': '0', 'STATISTICS_VALID_PERCENT': '100', 'STATISTICS_MEAN': '0', 'STATISTICS_APPROXIMATE': 'YES'}
fail
    line 820: did not get expected metadata
```

metzm commented:
All done.
Regarding tests, I modified TEST: stats_approx_nodata to also test for statistics stored in metadata: STATISTICS_VALID_PERCENT must be 50.
Regarding the helper function, I added SetValidPercent( GUIntBig nSampleCount, GUIntBig nValidCount ) to the class GDALRasterBand in gdal_priv.h, but I have no idea if this is the right place. Alternatively, something like SetValidPercentInternal( GDALRasterBand *poBand, GUIntBig nSampleCount, GUIntBig nValidCount ) could be used instead, but I am not convinced of this either.

rouault commented on gdal/gcore/gdal_priv.h:
```diff
@@ -1181,6 +1181,7 @@ class CPL_DLL GDALRasterBand : public GDALMajorObject
     virtual CPLErr SetStatistics( double dfMin, double dfMax,
                                   double dfMean, double dfStdDev );
     virtual CPLErr ComputeRasterMinMax( int, double* );
+    virtual void SetValidPercent( GUIntBig nSampleCount, GUIntBig nValidCount );
```
Please move it in the private: section of the class definition, and remove the virtual keyword

metzm replied:
Done

rouault commented:
Still issues with tabulation characters in gcore/gdalrasterband.cpp

metzm commented:
tabs have been removed and indentation has been adjusted in gcore/gdalrasterband.cpp. Sorry for the noise!

rouault commented:
Perfect. I've squashed the commits in a single one
